src/query/subquery_cache.c

```diff
@@ -0,0 +1,724 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * subquery_cache.c - Correlated Scalar Subquery Result Cache.
+ */
+
+
+#ident "$Id$"
+
+#include <stdio.h>
+#include <string.h>
+
+#include "xasl.h"
+#include "dbtype.h"
+#include "query_executor.h"
+#include "xasl_predicate.hpp"
+#include "regu_var.hpp"
+#include "object_representation.h"
+#include "system_parameter.h"
+
+#include "list_file.h"
+#include "db_set_function.h"
+
+#include "subquery_cache.h"
+
+typedef union _sq_regu_value
+{
+  /* fields used by both XASL interpreter and regulator */
+  DB_VALUE *dbvalptr;		/* for constant values */
+  QFILE_SORTED_LIST_ID *srlist_id;	/* sorted list identifier for subquery results */
+} sq_regu_value;
+
+typedef struct _sq_key
+{
+  DB_VALUE *pred_set;
+} sq_key;
+
+typedef struct _sq_val
+{
+  sq_regu_value val;
+  REGU_DATATYPE t;
+} sq_val;
+
+/**************************************************************************************/
+
+/* Static functions for sq_cache hash table. */
+
+static sq_key *sq_make_key (xasl_node * xasl);
+static sq_val *sq_make_val (REGU_VARIABLE * val);
+static void sq_free_key (sq_key * key);
+static void sq_free_val (sq_val * val);
+static void sq_unpack_val (sq_val * val, REGU_VARIABLE * retp);
+
+static unsigned int sq_hash_func (const void *key, unsigned int ht_size);
+static int sq_cmp_func (const void *key1, const void *key2);
+static int sq_rem_func (const void *key, void *data, void *args);
+
+static int sq_walk_xasl_check_not_caching (xasl_node * xasl);
+static int sq_walk_xasl_and_add_val_to_set (void *p, int type, sq_key * pred_set);
```

Comment:
sq_walk_xasl_check_not_caching ()
sq_walk_xasl_and_add_val_to_set ()
You need to move it to XASL generation in next step.